Failed 1 test when running `pytest`: `test_make_transparent - AssertionError: assert False`

1. Forked the repo and `git clone` to my local
2. Installed dependencies:
```
uv pip install -e '.[dev]'
Using Python 3.13.5 environment at: venv
Resolved 2 packages in 646ms
      Built py-image-border @ file:///Users/john_j_liu/GitHub/py-image-border
Prepared 2 packages in 489ms
Installed 2 packages in 3ms
 + pillow==11.3.0
 + py-image-border==2023.4.14.2
...
```
3. Also installed dev dependencies:
```
uv pip install pytest-cov 
Using Python 3.13.5 environment at: venv
Resolved 7 packages in 338ms
Installed 7 packages in 15ms
 + coverage==7.10.3
 + iniconfig==2.1.0
 + packaging==25.0
 + pluggy==1.6.0
 + pygments==2.19.2
 + pytest==8.4.1
 + pytest-cov==6.2.1
```
4. Run `pytest` and got 1 failure:
```
pytest
...
platform darwin -- Python 3.13.5, pytest-8.4.1, pluggy-1.6.0
...
collected 9 items                                                                                                                                           

tests/test_basic_functionality.py .....F...  
...
    def test_make_transparent():
        """Test that making the background transparent works."""
        # Note: currently, modified image ends up in same dir as source image.
        path_source = Path("tests/source_images/gh_screenshot.png")
        path_modified = Path("tests/source_images/gh_screenshot_bordered.png")
        path_reference = Path("tests/reference_images/gh_screenshot_transparent.png")
    
        cmd = f"python -m py_image_border.add_border {path_source} 0 --make-transparent"
        cmd_parts = cmd.split()
    
        subprocess.run(cmd_parts)
>       assert filecmp.cmp(path_modified, path_reference)
E       AssertionError: assert False
E        +  where False = <function cmp at 0x104567880>(PosixPath('tests/source_images/gh_screenshot_bordered.png'), PosixPath('tests/reference_images/gh_screenshot_transparent.png'))
E        +    where <function cmp at 0x104567880> = filecmp.cmp

tests/test_basic_functionality.py:91: AssertionError
...
```
5. Debug the test failure:
```
ll tests/source_images/gh_screenshot_bordered.png
-rw-r--r--  1 john_j_liu  staff    34K Aug 12 16:01 tests/source_images/gh_screenshot_bordered.png

ll tests/reference_images/gh_screenshot_transparent.png
-rw-r--r--  1 john_j_liu  staff    39K Aug 12 12:00 tests/reference_images/gh_screenshot_transparent.png
```
6. **Reasoning**: the metadata or compression can make PNG files binary different even if they look the same. We should use `Pillow` library to do pixel equality comparison.

--- a/tests/test_basic_functionality.py
+++ b/tests/test_basic_functionality.py
@@ -1,7 +1,71 @@
 """Initial tests of basic functionality."""
 
-import subprocess, filecmp
+from datetime import datetime
+from os.path import getsize, getmtime
 from pathlib import Path
+import filecmp, json, subprocess
+
+from PIL import Image, ImageChops
+
+
+TS_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+# --- Helper used in the tests. ---
+
+def readable_file_size(in_bytes: int) -> str:
+    """Convert bytes to human-readable KB, MB, or GB
+
+    Args:
+        in_bytes: input bytes
+
+    Returns:
+        str
+    """
+    for unit in ["B", "KB", "MB", "GB"]:
+        if in_bytes < 1024:
+            break
+        in_bytes /= 1024
+    return f"{in_bytes:.2f} {unit}"
+
+
+def are_images_equal(path1: Path | str, path2: Path | str) -> bool:
+    """Check if two image files are visually equal pixel-wise
+
+    Args:
+        path1: image1 file path
+        path2: image2 file path
+
+    Returns:
+        bool: True - visually equal, False - not visually equal
+    """
+    size1 = getsize(path1)
+    m_ts1 = datetime.fromtimestamp(getmtime(path1)).strftime(TS_FORMAT)
+    with Image.open(path1) as img1:
+        meta1 = {
+            "file_size": readable_file_size(size1),
+            "file_ts": m_ts1,
+            "format": img1.format,
+            "size": img1.size,
+            "mode": img1.mode,
+            "info": img1.info,
+        }
+        data1 = img1.convert("RGB")
+    size2 = getsize(path2)
+    m_ts2 = datetime.fromtimestamp(getmtime(path2)).strftime(TS_FORMAT)
+    with Image.open(path2) as img2:
+        meta2 = {
+            "file_size": readable_file_size(size2),
+            "file_ts": m_ts2,
+            "format": img2.format,
+            "size": img2.size,
+            "mode": img2.mode,
+            "info": img2.info,
+        }
+        data2 = img2.convert("RGB")
+    print(f"Meta of {path1}:\n{json.dumps(meta1, indent=2)}")
+    print(f"Meta of {path2}:\n{json.dumps(meta2, indent=2)}")
+    return ImageChops.difference(data1, data2).getbbox() is None
+
 
 # --- Tests for correct usage. ---
 
@@ -91,14 +155,7 @@ def test_make_transparent():
     # assert filecmp.cmp(path_modified, path_reference)
     # JL 2025-08-12: The metadata or compression can make PNG files binary different
     # even if they look the same. Fix: use Pillow library to compare pixel equality.
-    from PIL import Image, ImageChops
-
-    def images_are_equal(img1_path, img2_path):
-        img1 = Image.open(img1_path).convert("RGB")
-        img2 = Image.open(img2_path).convert("RGB")
-        return ImageChops.difference(img1, img2).getbbox() is None
-
-    assert images_are_equal(path_modified, path_reference)
+    assert are_images_equal(path_modified, path_reference) is True
 
 
 # --- Tests for incorrect usage. ---

--- a/tests/test_basic_functionality.py
+++ b/tests/test_basic_functionality.py
@@ -89,9 +89,8 @@ def test_make_transparent():
 
     subprocess.run(cmd_parts)
     # assert filecmp.cmp(path_modified, path_reference)
-    # JL 2025-08-12: Fix: the metadata or compression can make PNG files binary
-    # different even if they look the same. We should use Pillow library to do pixel
-    # equality comparison.
+    # JL 2025-08-12: The metadata or compression can make PNG files binary different
+    # even if they look the same. Fix: use Pillow library to compare pixel equality.
     from PIL import Image, ImageChops
 
     def images_are_equal(img1_path, img2_path):

--- a/tests/test_basic_functionality.py
+++ b/tests/test_basic_functionality.py
@@ -88,7 +88,18 @@ def test_make_transparent():
     cmd_parts = cmd.split()
 
     subprocess.run(cmd_parts)
-    assert filecmp.cmp(path_modified, path_reference)
+    # assert filecmp.cmp(path_modified, path_reference)
+    # JL 2025-08-12: Fix: the metadata or compression can make PNG files binary
+    # different even if they look the same. We should use Pillow library to do pixel
+    # equality comparison.
+    from PIL import Image, ImageChops
+
+    def images_are_equal(img1_path, img2_path):
+        img1 = Image.open(img1_path).convert("RGB")
+        img2 = Image.open(img2_path).convert("RGB")
+        return ImageChops.difference(img1, img2).getbbox() is None
+
+    return images_are_equal(path_modified, path_reference)
 
 
 # --- Tests for incorrect usage. ---

--- a/tests/test_basic_functionality.py
+++ b/tests/test_basic_functionality.py
@@ -99,7 +99,7 @@ def test_make_transparent():
         img2 = Image.open(img2_path).convert("RGB")
         return ImageChops.difference(img1, img2).getbbox() is None
 
-    return images_are_equal(path_modified, path_reference)
+    assert images_are_equal(path_modified, path_reference)
 
 
 # --- Tests for incorrect usage. ---

